epochStart/bootstrap/shardStorageHandler.go

```diff
@@ -882,34 +882,70 @@ func (ssh *shardStorageHandler) saveTriggerRegistry(components *ComponentsNeeded
 	return bootstrapKey, nil
 }
 
-func getMiniBlockHeadersForDest(metaBlock *block.MetaBlock, destId uint32) map[string]block.MiniBlockHeader {
+func getMiniBlockHeadersForDest(metaBlock data.MetaHeaderHandler, destId uint32) map[string]block.MiniBlockHeader {
 	hashDst := make(map[string]block.MiniBlockHeader)
-	for i := 0; i < len(metaBlock.ShardInfo); i++ {
-		if metaBlock.ShardInfo[i].ShardID == destId {
+	for i := 0; i < len(metaBlock.GetShardInfoHandlers()); i++ {
+		if metaBlock.GetShardInfoHandlers()[i].GetShardID() == destId {
 			continue
 		}
 
-		for _, val := range metaBlock.ShardInfo[i].ShardMiniBlockHeaders {
-			isCrossShardDestMe := val.ReceiverShardID == destId && val.SenderShardID != destId
+		for _, val := range metaBlock.GetShardInfoHandlers()[i].GetShardMiniBlockHeaderHandlers() {
+			isCrossShardDestMe := val.GetReceiverShardID() == destId && val.GetSenderShardID() != destId
 			if !isCrossShardDestMe {
 				continue
 			}
 
-			hashDst[string(val.Hash)] = val
+			miniBlockHeader, ok := val.(*block.MiniBlockHeader)
+			if !ok {
+				log.Warn("wrong type assertion for mini block header handler", "err", epochStart.ErrWrongTypeAssertion)
+				continue
+			}
+			hashDst[string(val.GetHash())] = *miniBlockHeader
 		}
 	}
 
-	for _, val := range metaBlock.MiniBlockHeaders {
-		isCrossShardDestMe := (val.ReceiverShardID == destId || val.ReceiverShardID == core.AllShardId) && val.SenderShardID != destId
+	miniBlockHandlers := getMetaHeaderMiniBlockHandlersFromExecutionResults(metaBlock)
+
+	for _, val := range miniBlockHandlers {
+		isCrossShardDestMe := (val.GetReceiverShardID() == destId || val.GetReceiverShardID() == core.AllShardId) && val.GetSenderShardID() != destId
 		if !isCrossShardDestMe {
 			continue
 		}
-		hashDst[string(val.Hash)] = val
+
+		miniBlockHeader, ok := val.(*block.MiniBlockHeader)
+		if !ok {
+			log.Warn("wrong type assertion for mini block header handler", "err", epochStart.ErrWrongTypeAssertion)
+			continue
+		}
+		hashDst[string(val.GetHash())] = *miniBlockHeader
 	}
 
 	return hashDst
 }
 
+func getMetaHeaderMiniBlockHandlersFromExecutionResults(
+	metaBlock data.MetaHeaderHandler,
+) []data.MiniBlockHeaderHandler {
+	miniBlockHeaderHandlers := metaBlock.GetMiniBlockHeaderHandlers()
+	if !metaBlock.IsHeaderV3() {
+		return miniBlockHeaderHandlers
+	}
+
+	lastExecResult, err := common.GetLastBaseExecutionResultHandler(metaBlock)
```

Comment:
can you add some unit tests for this function?

The common.GetLastBaseExecutionResultHandler func should only return the summary of the last execution result, which does not contain the miniblock headers. It also can be from a previous block if the current block did not add any new results.

this means that on line 940 you will have a type assertion error.

you probably need to iterate instead over the header.GetExecutionResultsHandlers() and extract from all (if any) the miniblock header handlers, as you do in baseProcess.go L2515